src/libstore/build.cc

```diff
@@ -3673,7 +3673,7 @@ void DerivationGoal::registerOutputs()
 
                 Path actualDest = worker.store.toRealPath(worker.store.printStorePath(dest));
 
-                if (worker.store.isValidPath(dest))
+                if (worker.store.isValidPath(dest) || buildMode == bmCheck)
```

Comment:
Why that condition is here at all? Feels like if the hashes don't match, the builder should error. What am I missing?

Reply:
I believe the point here is just delaying the exception -- the exception is re-thrown at [the end of the function](https://github.com/NixOS/nix/pull/3432/files#diff-267a6391980cbd9a743958945ff44ef9R3842). Unfortunately not in the case when `buildMode == bmCheck`.